### ISSUE TITLE
rsyslog warns about environment variable TZ not being set

### ISSUE THREAD

**Issue by mbiebl** [rsyslog/rsyslog#2994]
Version: 8.37.0

When starting rsyslog, I get the following warning message:
```
Sep 05 16:39:29 pluto rsyslogd[689]: environment variable TZ is not set, auto correcting this to TZ=/etc/localtime  
```

I think this is misleading. There is nothing wrong if TZ is not set. I would even argue that it's much more common that TZ is *not* set, on Debian/Ubuntu it isn't.
It's alarming to see this warning message on basically every Debian/Ubuntu machine (not sure if other distros set the TZ env var by default or not).

I would suggest downgrading this message to debug or reversing the logic. I.e. only log something if TZ *is* set and log nothing if TZ is not set and /etc/timezone is used.

**Comment by mbiebl** [comment]
when running systemctl status rsyslog, this is how it looks like:
```
$ systemctl status rsyslog
● rsyslog.service - System Logging Service
   Loaded: loaded (/etc/systemd/system/rsyslog.service; enabled; vendor preset: enabled)
   Active: active (running) since Wed 2018-09-05 16:39:29 CEST; 3h 49min ago
     Docs: man:rsyslogd(8)
           http://www.rsyslog.com/doc/
 Main PID: 689 (rsyslogd)
    Tasks: 4 (limit: 4915)
   Memory: 7.3M
   CGroup: /system.slice/rsyslog.service
           └─689 /usr/sbin/rsyslogd -n -iNONE

Sep 05 16:39:29 pluto systemd[1]: Starting System Logging Service...
Sep 05 16:39:29 pluto rsyslogd[689]: environment variable TZ is not set, auto correcting this to TZ=/etc/localtime  
Sep 05 16:39:29 pluto rsyslogd[689]: imuxsock: Acquired UNIX socket '/run/systemd/journal/syslog' (fd 3) from system
Sep 05 16:39:29 pluto systemd[1]: Started System Logging Service.
Sep 05 16:39:29 pluto rsyslogd[689]:  [origin software="rsyslogd" swVersion="8.37.0" x-pid="689" x-info="http://www.
```

**Comment by mbiebl** [comment]
I might be mistaken, but I thought glibc automatically consults /etc/localtime. E.g. if I run date via strace, I see
```
openat(AT_FDCWD, "/etc/localtime", O_RDONLY|O_CLOEXEC) = 3
```

**Comment by mbiebl** [comment]
And quoting https://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html
```
In POSIX systems, a user can specify the time zone by means of the TZ environment variable. For information about how to set environment variables, see Environment Variables. The functions for accessing the time zone are declared in time.h.

You should not normally need to set TZ. If the system is configured properly, the default time zone will be correct. You might set TZ if you are using a computer over a network from a different time zone, and would like times reported to you in the time zone local to you, rather than what is local to the computer.
```

**Comment by mbiebl** [comment]
And later on:
```
If the TZ environment variable does not have a value, the operation chooses a time zone by default. In 
the GNU C Library, the default time zone is like the specification ‘TZ=:/etc/localtime’ (or 
‘TZ=:/usr/local/etc/localtime’, depending on how the GNU C Library was configured; see Installation). 
Other C libraries use their own rule for choosing the default time zone, so there is little we can say about 
them.
```

**Comment by rgerhards** [comment]
This was the issue:

https://github.com/rsyslog/rsyslog/issues/2054

There is quite some info in it.

**Comment by mbiebl** [comment]
I still see no good justification why a missing TZ env var is considered an error if /etc/localtime is properly setup.

**Comment by rgerhards** [comment]
@jpvriel can you comment?

**Comment by JPvRiel** [comment]
> I still see no good justification why a missing TZ env var is considered an error if /etc/localtime is properly setup.

I noticed the bad behaviour with TZ not being set when rsyslog was run within a container (built from a CentOS base). One nuanced case about the previous issue I logged (#2054) is that perhaps @mbiebl is correct and most distro's do the correct setup in /etc/localtime? However containers are a corner case where this assumption would be wrong?

I tried this to illustrate the potential issue with containers as follows:

```
$ docker run -it centos zdump /etc/localtime
/etc/localtime  Mon Sep 10 13:19:32 2018 UTC
$ zdump /etc/localtime 
/etc/localtime  Mon Sep 10 15:19:40 2018 SAST
```

And here's the affect if one does supply TZ to a container:

```
$ docker run -e TZ=Africa/Johannesburg -it centos date
Mon Sep 10 15:23:52 SAST 2018
$ docker run -it centos date
Mon Sep 10 13:24:00 UTC 2018
```

**Comment by JPvRiel** [comment]
@mbiebl, other than the risk of assuming the wrong timezone within a container, which is bad when rsyslog relays messages recived with RFC3164 (deficient old syslog format where headers don't include the timeszone), there is also the potential performance issues. [How setting the TZ environment variable avoids thousands of system calls](https://blog.packagecloud.io/eng/2017/02/21/set-environment-variable-save-thousands-of-system-calls/) is the referance that made me concerned about TZ not being set. Not sure if rsyslog would be impacted by this and if recent versions of glibc still have this performance issue, but, for example, the post did claim it affected Ubuntu LTS 16.04 which will still be officially supported for more than 2 years.

Given the above, my view is it's fair for rsyslog to at least issue a startup notice about TZ not being set. Setting TZ is safer, unless someone wants to dig about and find out under which contexts rsyslog can trust that glibc won't run unnecessary time system calls and can trust /etc/localtime is correct (one might argue containers are bad at this, and fail to set /etc/localtime properly).

**Comment by mbiebl** [comment]
huh, you've setup /etc/localtime in the container to be UTC so it's rather obvious that the timezone inside the container would be UTC. If you want to override that to match the host system's timezone, then that's fine. What I'm arguing about is, that rsyslog should *not* warn if TZ is not set, as this is perfectly normal.

**Comment by rgerhards** [comment]
> that rsyslog should not warn if TZ is not set,

Actually, the warning is less that TZ is not set, but more to the fact that rsyslog sets it to a specific value. If the value we use is incorrect, problems will probably occur. And the performance implications @jpvriel quoted made me think it is worth setting TZ.

Nevertheless, I can make the warning optional with default "off". So it can be turned on if strange things happen. From what I read in this issue, the occurrence of such strange things looks pretty unlikely (maybe the default should be different on non-Linux, where I think it still is common to use TZ).

**Comment by JPvRiel** [comment]
> you've setup /etc/localtime in the container to be UTC so it's rather obvious that the timezone inside the container would be UTC

I didn't explicitly set anything. That's the default for docker at least... Docker engine containers don't match your assumption that /etc/localtime is correctly set by default. Either expect the docker container defaults of UTC (which as I explained is actually bad because of how rsyslog interprets RFC3164 message time) or explicitly set either the /etc/localtime or TZ for a container. TZ is probably simpler way given container patterns seem to expect being controlled via env vars instead of config files (which need more cumbersome setup with things like baked into the image Dockerfile, bind mounts or config file volumes, etc).

For anyone interested, this gets messy: [Docker Container time & timezone (will not reflect changes)](https://serverfault.com/questions/683605/docker-container-time-timezone-will-not-reflect-changes)

> rsyslog should not warn if TZ is not set, as this is perfectly normal.

As I tried to explain, this might be normal for traditional linux native host processes, but has bad consequences if run inside a docker container and may have a negative performance impact. So even if normal, not optimal and can cause bad time skew bugs with message timezone interpretation inside containers.

**Comment by mbiebl** [comment]
Are you really running rsyslog inside a docker container? Isn't that "special"?

**Comment by JPvRiel** [comment]
> Are you really running rsyslog inside a docker container? Isn't that "special"?

Fair point. It's more 'bleeding-edge', but not entirely uncommon if I observe dockerhub repos doing the same and rsyslog is experimenting with supporting this in future [rsyslog-docker](https://github.com/rsyslog/rsyslog-docker), so it's likely to become a more normal use-case.
